lib/twom.c

```diff
@@ -0,0 +1,3312 @@
+                          " parent_key=<%.*s> parent_offset=<%08llX)",
+                          db->fname, (int)KEYLEN(nextptr), KEYPTR(nextptr), (LLU)next[0],
+                          (int)KEYLEN(aptr), KEYPTR(aptr), (LLU)ancestor);
+                return TWOM_IOERROR;
+            }
+            dirty_size += RECLEN(aptr);
+            ancestor = ANCESTOR(aptr);
+        }
+
+        uint8_t level = LEVEL(nextptr);
+        for (i = 1; i < level; i++) {
+            /* check the old pointer was to here */
+            if (next[i] != next[0]) {
+                db->error("broken linkage for consistent",
+                          "fname=<%s> offset=<%08llX> level=<%d>"
+                          " expected=<%08llX>",
+                          db->fname, (LLU)next[0], i, (LLU)next[i]);
+                return TWOM_IOERROR;
+            }
+            /* and advance to the new pointer */
+            next[i] = NEXTN(nextptr, i);
+        }
+        next[0] = advance0(nextptr);
+
+        // count if record or tombstone
+        if (deleted_offset) dirty_size += RECLEN(nextptr);
+        else num_records++;
+
+        ptr = nextptr;
+    }
+
+    for (i = 0; i < MAXLEVEL; i++) {
+        if (next[i]) {
+            db->error("broken tail for consistent",
+                      "filename=<%s> offset=<%08llX> level=<%d>",
+                      db->fname, (LLU)next[i], i);
+            return TWOM_IOERROR;
+        }
+    }
+
+    /* we walked the whole file and saw every pointer */
+
+    if (num_records != txn->file->header.num_records) {
+        db->error("record count mismatch for consistent",
+                  "filename=<%s> num_records=<%llu> expected_records=<%llu>",
+                  db->fname, (LLU)num_records, (LLU)txn->file->header.num_records);
+        return TWOM_IOERROR;
+    }
+
+    if (dirty_size != txn->file->header.dirty_size) {
+        db->error("dirty_size mismatch for consistent",
+                  "filename=<%s> dirty_size=<%llu> expected_size=<%llu>",
+                  db->fname, (LLU)dirty_size, (LLU)txn->file->header.dirty_size);
+        return TWOM_IOERROR;
+    }
+
+    return 0;
+}
+
+/*********************** PUBLIC API ******************************/
+
+// we refcount the database object by filename, so if you open twice in
+// the same process, we don't have flock shenanigans, because flock state
+// is process-wise, not file-descriptor specific, so it's not safe to just
+// open the file in separate objects and lock it separately.
+int twom_db_open(const char *fname, struct twom_open_data *setup,
+                 struct twom_db **ret, struct twom_txn **txnp)
+{
+    struct twom_db *mydb;
+    int r = 0;
+
+    /* do we already have this DB open? */
+    for (mydb = open_twom; mydb; mydb = mydb->next) {
+        if (strcmp(mydb->fname, fname)) continue;
+        if (txnp) {
+            r = twom_db_begin_txn(mydb, setup->flags & TWOM_SHARED, txnp);
```

Comment:
I want to clarify the purpose of thatsetup->flags & TWOM_SHARED: This deliberately removes all flags, only keeping TWOM_SHARED if that's what the caller passed as flags. Flags such as NOSYNC get wiped, presumably because a second open on an already open database should not reset that flag.
Isn't that then kind of lying to the caller, e.g. if I write the code to expect NOSYNC being enabled, but the db already got opened without NOSYNC in the first open?

Reply:
Yeah, that was a mistake - this USED to take a boolean.  I've switched it to pass the whole flags through.

It might be useful to check flags and return a "you're holding it wrong" error.  I'll add a comment there, but otherwise leave it for now.